bazel/cc_proto_descriptor_library/file_descriptor_generator.cc

```diff
@@ -0,0 +1,180 @@
+    return stripped_name;
+  } else if (absl::ConsumeSuffix(&stripped_name, ".protodevel")) {
+    return stripped_name;
+  }
+  return stripped_name;
+}
+
+std::string getDescriptorHeaderName(const google::protobuf::FileDescriptor* file) {
+  return absl::StrCat(getFileBaseName(file), "_descriptor.pb.h");
+}
+
+std::string getDescriptorSourceName(const google::protobuf::FileDescriptor* file) {
+  return absl::StrCat(getFileBaseName(file), "_descriptor.pb.cc");
+}
+
+std::string getDescriptorNamespace(const google::protobuf::FileDescriptor* file) {
+  return absl::AsciiStrToLower(
+      absl::StrReplaceAll(getFileBaseName(file), {{"/", "_"}, {"-", "_"}}));
+}
+
+std::string getDependencyFileDescriptorInfoSymbol(const google::protobuf::FileDescriptor* file) {
+  return absl::StrFormat("%s::kFileDescriptorInfo", getDescriptorNamespace(file));
+}
+
+std::string getDependencyFileDescriptorHeaderGuard(const google::protobuf::FileDescriptor* file) {
+  std::string header_path = getDescriptorHeaderName(file);
+  return absl::AsciiStrToUpper(
+      absl::StrReplaceAll(header_path, {{"/", "_"}, {".", "_"}, {"-", "_"}}));
+}
+
+bool generateHeader(const google::protobuf::FileDescriptor* file,
+                    google::protobuf::io::ZeroCopyOutputStream* output_stream) {
+  auto header_guard = getDependencyFileDescriptorHeaderGuard(file);
+  auto unique_namespace = getDescriptorNamespace(file);
+
+  std::stringstream contents;
+  contents << absl::StrFormat(R"text(
+#ifndef %s
+#define %s
+
+#include "absl/base/attributes.h"
+
+namespace cc_proto_descriptor_library {
+namespace internal {
+
+struct FileDescriptorInfo;
+
+} // namespace internal
+} // namespace cc_proto_descriptor_library
+
+namespace protobuf {
+namespace reflection {
+namespace %s {
+
+extern const ::cc_proto_descriptor_library::internal::FileDescriptorInfo kFileDescriptorInfo;
+} // namespace %s
+} // namespace reflection
+} // namespace protobuf
+
+#endif // %s
+
+)text",
+                              header_guard, header_guard, unique_namespace, unique_namespace,
+                              header_guard);
+
+  google::protobuf::io::CodedOutputStream output(output_stream);
+  output.WriteString(contents.str());
+  output.Trim();
+  return !output.HadError();
+}
+
+bool generateSource(const google::protobuf::FileDescriptor* file,
+                    google::protobuf::io::ZeroCopyOutputStream* output_stream) {
+  auto unique_namespace = getDescriptorNamespace(file);
+  std::stringstream contents;
+
+  contents << absl::StrFormat("#include \"%s\"\n", getDescriptorHeaderName(file));
+  for (int i = 0; i < file->dependency_count(); ++i) {
+    contents << absl::StrFormat("#include \"%s\"\n", getDescriptorHeaderName(file->dependency(i)));
+  }
+  contents << R"text(#include "bazel/cc_proto_descriptor_library/file_descriptor_info.h"
+)text";
+
+  google::protobuf::FileDescriptorProto file_descriptor_proto;
+  file->CopyTo(&file_descriptor_proto);
```

Comment:
in google3, this seems to call `proto2::compiler::StripSourceRetentionOptions` instead, not sure I understand the difference in what's happening here?  but if it works, i trust it :)

Reply:
Oh, right. That's in the code commented out below. Now that we're using 23.1 I can use that. Thanks for the reminder!